### ISSUE TITLE
machined: "machinectl copy-from" and "machinectl copy-to" need to adjust uid/gid accordingl to userns if containers uses them

### ISSUE THREAD

**Issue by poettering** [systemd/systemd#4078]
Subject says it all...

See: https://lists.freedesktop.org/archives/systemd-devel/2016-September/037418.html

**Comment by gabx** [comment]
Before this issue is solved, is there any temporary hack which would allow me to use my container? For now, because of all these wrong uid/gid, I can't use web services from the container.

**Comment by evverx** [comment]
> is there any temporary hack which would allow me to use my container? 
- install/[build](https://github.com/lxc/lxd#building-the-tools) the [`fuidshift`](https://github.com/lxc/lxd/tree/master/fuidshift)
- detect the `uid/gid` of the container

```
-bash-4.3# ls -ld /var/lib/machines/your-cont/
drwxr-xr-x. 13 325124096 325124096 1024 Sep 24 13:58 /var/lib/machines/your-cont/
```
- copy something to the container

```
cp something /var/lib/machines/your-cont
```
- fix permissions

```
fuidshift /var/lib/machines/your-cont/something b:0:325124096:65536
```

Note: `fuidshift` doesn't handle acls

**Comment by gabx** [comment]
Thank you. In fact, I have done something similar, as on the host I see owner:group as numeric. Example:

`# ls -al /var/lib/machines/poppy/
dr-xr-xr-x 1 363397120 363397120    0 Feb  3  2016 boot/
drwxrwxr-x 1 363397120 363397120   62 Aug 26 19:59 db/
drwxr-xr-x 1 363397120 363397120    8 Apr 27  2015 dev/`

Back to my container, when I wan to chown from nobody:nobody (most common error) to root:root, I just run:
`# chown -R 363397120:363397120 MyFolder`

For now, it works, and I am able to run my container properly. On my container, I am left with very few nobody:nobody. /proc being the only one I can't change from nobody:nobody.
I will install fuidshif and see.
TY

**Comment by gabx** [comment]
On Sun, Sep 25, 2016 at 11:08 PM Evgeny Vereshchagin <
notifications@github.com> wrote:

is there any temporary hack which would allow me to use my container?

> - install/build https://github.com/lxc/lxd#building-the-tools the
>   fuidshift https://github.com/lxc/lxd/tree/master/fuidshift
> - detect the uid/gid of the container
> 
> -bash-4.3# ls -ld /var/lib/machines/your-cont/
> drwxr-xr-x. 13 325124096 325124096 1024 Sep 24 13:58 /var/lib/machines/your-cont/
> - copy something to the container
> 
> cp something /var/lib/machines/your-cont
> - fix permissions
> 
> fuidshift /var/lib/machines/your-cont/something b:0:325124096:65536
> 
> Note: fuidshift doesn't handle acls
> 
> I built fuidshift and gave a try. The result was the expected one. This
> method asks less as I can do everything from the host, when I had to write
> on both with my trick.

Now please can you tell me where the 65536 comes from, as well as the 0 ?
From $ fuidshift --help, I can't see any good explanation.

Thank you

**Comment by evverx** [comment]
https://www.freedesktop.org/software/systemd/man/systemd-nspawn.html#--private-users=

> The value "pick" turns on user namespacing. In this case the UID/GID range is automatically chosen. As first step, the file owner of the root directory of the container's directory tree is read, and it is checked that it is currently not used by the system otherwise (in particular, that no other container is using it). If this check is successful, the UID/GID range determined this way is used, similar to the behaviour if "yes" is specified. If the check is not successful (and thus the UID/GID range indicated in the root directory's file owner is already used elsewhere) a new – currently unused – UID/GID range of 65536 UIDs/GIDs is randomly chosen between the host UID/GIDs of 524288 and 1878982656, always starting at a multiple of 65536. This setting implies --private-users-chown (see below), which has the effect that the files and directories in the container's directory tree will be owned by the appropriate users of the range picked. Using this option makes user namespace behaviour fully automatic. Note that the first invocation of a previously unused container image might result in picking a new UID/GID range for it, and thus in the (possibly expensive) file ownership adjustment operation. However, subsequent invocations of the container will be cheap (unless of course the picked UID/GID range is assigned to a different use by then).

https://lwn.net/Articles/532593/
https://lwn.net/Articles/540087/
http://man7.org/linux/man-pages/man7/user_namespaces.7.html

**Comment by gabx** [comment]
On Sun, Sep 25, 2016 at 11:08 PM Evgeny Vereshchagin <
notifications@github.com> wrote:

> is there any temporary hack which would allow me to use my container?
> - install/build https://github.com/lxc/lxd#building-the-tools the
>   fuidshift https://github.com/lxc/lxd/tree/master/fuidshift
> - detect the uid/gid of the container
> 
> -bash-4.3# ls -ld /var/lib/machines/your-cont/
> drwxr-xr-x. 13 325124096 325124096 1024 Sep 24 13:58 /var/lib/machines/your-cont/
> - copy something to the container
> 
> cp something /var/lib/machines/your-cont
> - fix permissions
> 
> fuidshift /var/lib/machines/your-cont/something b:0:325124096:65536
> 
> Note: fuidshift doesn't handle acls

I have been able to fix mostly everything, but I am still left with a
broken owner for files and directory of /proc on the container.

From the host:

# ls -al /var/lib/machines/poppy/

total 20K
dr-xr-xr-x 1 363397120 363397120  198 Oct  4 15:54 ./
drwx------ 3 root      root      4.0K Oct  1 15:53 ../
dr-xr-xr-x 1 363397120 363397120    0 Feb  3  2016 boot/
drwxrwxr-x 1 363397120 363397120   62 Aug 26 19:59 db/
drwxr-xr-x 1 363397120 363397120    8 Apr 27  2015 dev/
drwxr-xr-x 1 363397120 363397120 4.1K Feb  3  2016 etc/
drwxr-xr-x 1 363397120 363397120   76 Feb  3  2016 home/
drwxrwxrwx 1 363397120 363397120    0 Sep 23 12:39 keybase/
drwxr-xr-x 1 363397120 363397120    0 Feb  3  2016 media/
drwxr-xr-x 1 363397120 363397120    0 Feb  3  2016 mnt/
drwxr-xr-x 1 363397120 363397120   56 Feb  3  2016 opt/
dr-xr-xr-x 1 363397120 363397120    0 Aug 16  2014 proc/
dr-xr-x--- 1 363397120 363397120  378 Sep 25 11:15 root/
drwxr-xr-x 1 363397120 363397120  140 Apr 27  2015 run/
drwxr-xr-x 1 363397120 363397120    6 Mar  3  2016 share/
drwxr-xr-x 1 363397120 363397120    0 Feb  3  2016 srv/
drwxrwxr-x 1 363397120 363397130  260 Sep 26 10:18 storage/
dr-xr-xr-x 1 363397120 363397120    0 Aug 16  2014 sys/
drwxrwxrwt 1 363397120 363397120   94 Apr 27  2015 tmp/
drwxr-xr-x 1 363397120 363397120  100 Dec 14  2015 usr/
drwxr-xr-x 1 363397120 363397120  194 Mar 19  2016 var/
-rw-r--r-- 1 363397120 363397120    0 Sep  8 13:57 .autorelabel
lrwxrwxrwx 1 363397120 363397120    7 Feb  3  2016 bin -> usr/bin/
lrwxrwxrwx 1 363397120 363397120    7 Feb  3  2016 lib -> usr/lib/
lrwxrwxrwx 1 363397120 363397120    9 Feb  3  2016 lib64 -> usr/lib64/
lrwxrwxrwx 1 363397120 363397120    8 Feb  3  2016 sbin -> usr/sbin/

Now, from the container:

$ ls -al /proc
................
 dr-xr-xr-x   9 redis           redis              0 Oct  1 15:53 66/
dr-xr-xr-x   9 root            root               0 Oct  1 15:53 72/
dr-xr-xr-x   9 root            root               0 Oct  1 15:53 75/
dr-xr-xr-x   9 opendkim        postfix            0 Oct  1 15:53 81/
dr-xr-xr-x   9 postgres        postgres           0 Oct  1 15:53 86/
dr-xr-xr-x   9 root            root               0 Oct  1 15:53 95/
dr-xr-xr-x   9 postfix         postfix            0 Oct  4 15:56 968/
dr-xr-xr-x   2 nobody          nobody             0 Oct  4 16:13 acpi/
dr-xr-xr-x   5 nobody          nobody             0 Oct  4 16:13 asound/
dr-xr-xr-x   4 nobody          nobody             0 Oct  4 16:13 bus/
dr-xr-xr-x   2 nobody          nobody             0 Oct  4 16:13 driver/
dr-xr-xr-x   4 nobody          nobody             0 Oct  4 16:13 fs/
dr-xr-xr-x  40 nobody          nobody             0 Oct  4 16:13 irq/
dr-xr-xr-x   3 nobody          nobody             0 Oct  4 16:13 scsi/
dr-xr-xr-x   1 nobody          nobody             0 Oct  1 15:53 sys/
dr-xr-xr-x   2 nobody          nobody             0 Oct  4 16:13 sysvipc/
dr-xr-xr-x   4 nobody          nobody             0 Oct  4 16:13 tty/
-r--r--r--   1 nobody          nobody             0 Oct  4 16:13 buddyinfo
-r--r--r--   1 nobody          nobody             0 Oct  4 16:13 cgroups
............................

As you can see, nobody:nobody is back. And this time, I have no idea how to
fix this issue.
TY again for your help

**Comment by gabx** [comment]
On Tue, Oct 4, 2016 at 4:16 PM arnaud gaboury arnaud.gaboury@gmail.com
wrote:

> On Sun, Sep 25, 2016 at 11:08 PM Evgeny Vereshchagin <
> notifications@github.com> wrote:
> 
> is there any temporary hack which would allow me to use my container?
> - install/build https://github.com/lxc/lxd#building-the-tools the
>   fuidshift https://github.com/lxc/lxd/tree/master/fuidshift
> - detect the uid/gid of the container
> 
> -bash-4.3# ls -ld /var/lib/machines/your-cont/
> drwxr-xr-x. 13 325124096 325124096 1024 Sep 24 13:58 /var/lib/machines/your-cont/
> - copy something to the container
> 
> cp something /var/lib/machines/your-cont
> - fix permissions
> 
> fuidshift /var/lib/machines/your-cont/something b:0:325124096:65536
> 
> Note: fuidshift doesn't handle acls
> 
> I have been able to fix mostly everything, but I am still left with a
> broken owner for files and directory of /proc on the container.
> 
> From the host:
> 
> # ls -al /var/lib/machines/poppy/
> 
> total 20K
> dr-xr-xr-x 1 363397120 363397120  198 Oct  4 15:54 ./
> drwx------ 3 root      root      4.0K Oct  1 15:53 ../
> dr-xr-xr-x 1 363397120 363397120    0 Feb  3  2016 boot/
> drwxrwxr-x 1 363397120 363397120   62 Aug 26 19:59 db/
> drwxr-xr-x 1 363397120 363397120    8 Apr 27  2015 dev/
> drwxr-xr-x 1 363397120 363397120 4.1K Feb  3  2016 etc/
> drwxr-xr-x 1 363397120 363397120   76 Feb  3  2016 home/
> drwxrwxrwx 1 363397120 363397120    0 Sep 23 12:39 keybase/
> drwxr-xr-x 1 363397120 363397120    0 Feb  3  2016 media/
> drwxr-xr-x 1 363397120 363397120    0 Feb  3  2016 mnt/
> drwxr-xr-x 1 363397120 363397120   56 Feb  3  2016 opt/
> dr-xr-xr-x 1 363397120 363397120    0 Aug 16  2014 proc/
> dr-xr-x--- 1 363397120 363397120  378 Sep 25 11:15 root/
> drwxr-xr-x 1 363397120 363397120  140 Apr 27  2015 run/
> drwxr-xr-x 1 363397120 363397120    6 Mar  3  2016 share/
> drwxr-xr-x 1 363397120 363397120    0 Feb  3  2016 srv/
> drwxrwxr-x 1 363397120 363397130  260 Sep 26 10:18 storage/
> dr-xr-xr-x 1 363397120 363397120    0 Aug 16  2014 sys/
> drwxrwxrwt 1 363397120 363397120   94 Apr 27  2015 tmp/
> drwxr-xr-x 1 363397120 363397120  100 Dec 14  2015 usr/
> drwxr-xr-x 1 363397120 363397120  194 Mar 19  2016 var/
> -rw-r--r-- 1 363397120 363397120    0 Sep  8 13:57 .autorelabel
> lrwxrwxrwx 1 363397120 363397120    7 Feb  3  2016 bin -> usr/bin/
> lrwxrwxrwx 1 363397120 363397120    7 Feb  3  2016 lib -> usr/lib/
> lrwxrwxrwx 1 363397120 363397120    9 Feb  3  2016 lib64 -> usr/lib64/
> lrwxrwxrwx 1 363397120 363397120    8 Feb  3  2016 sbin -> usr/sbin/
> 
> Now, from the container:
> 
> $ ls -al /proc
> ................
>  dr-xr-xr-x   9 redis           redis              0 Oct  1 15:53 66/
> dr-xr-xr-x   9 root            root               0 Oct  1 15:53 72/
> dr-xr-xr-x   9 root            root               0 Oct  1 15:53 75/
> dr-xr-xr-x   9 opendkim        postfix            0 Oct  1 15:53 81/
> dr-xr-xr-x   9 postgres        postgres           0 Oct  1 15:53 86/
> dr-xr-xr-x   9 root            root               0 Oct  1 15:53 95/
> dr-xr-xr-x   9 postfix         postfix            0 Oct  4 15:56 968/
> dr-xr-xr-x   2 nobody          nobody             0 Oct  4 16:13 acpi/
> dr-xr-xr-x   5 nobody          nobody             0 Oct  4 16:13 asound/
> dr-xr-xr-x   4 nobody          nobody             0 Oct  4 16:13 bus/
> dr-xr-xr-x   2 nobody          nobody             0 Oct  4 16:13 driver/
> dr-xr-xr-x   4 nobody          nobody             0 Oct  4 16:13 fs/
> dr-xr-xr-x  40 nobody          nobody             0 Oct  4 16:13 irq/
> dr-xr-xr-x   3 nobody          nobody             0 Oct  4 16:13 scsi/
> dr-xr-xr-x   1 nobody          nobody             0 Oct  1 15:53 sys/
> dr-xr-xr-x   2 nobody          nobody             0 Oct  4 16:13 sysvipc/
> dr-xr-xr-x   4 nobody          nobody             0 Oct  4 16:13 tty/
> -r--r--r--   1 nobody          nobody             0 Oct  4 16:13 buddyinfo
> -r--r--r--   1 nobody          nobody             0 Oct  4 16:13 cgroups
> ............................
> 
> As you can see, nobody:nobody is back. And this time, I have no idea how
> to fix this issue.
> TY again for your help

EDIT (sorry for this second post): I forgot to mention I don't have any
access from host to folders and files inside the /proc directory.

# ls -al /var/lib/machines/poppy/proc

 dr-xr-xr-x 1 363397120 363397120   0 Aug 16  2014 ./
dr-xr-xr-x 1 363397120 363397120 198 Oct  4 15:54 ../

return is empty.

**Comment by evverx** [comment]
> I have been able to fix mostly everything, but I am still left with a
>  broken owner for files and directory of /proc on the container.

@gabx, see https://github.com/systemd/systemd/pull/3093#issuecomment-214269187

**Comment by gabx** [comment]
On Tue, Oct 4, 2016 at 4:29 PM Evgeny Vereshchagin notifications@github.com
wrote:

> I have been able to fix mostly everything, but I am still left with a
> broken owner for files and directory of /proc on the container.
> 
> @gabx https://github.com/gabx, see #3093 (comment)
> https://github.com/systemd/systemd/pull/3093#issuecomment-214269187

OK, but now? I can't update my container filesystem. The package manager
(dnf in fedora) complains :
< error: unpacking of archive failed on file /proc: cpio: chown>

Do I have any solution?

TY.

**Comment by evverx** [comment]
I'm not sure I understand what are you trying to do?
Do you run `dnf install something` inside the container?

**Comment by gabx** [comment]
On Tue, Oct 4, 2016 at 4:52 PM Evgeny Vereshchagin notifications@github.com
wrote:

> I'm not sure I understand what are you trying to do?
> Do you run dnf install something inside the container?

I am trying to update my Fedora container (running inside Archlinux).  dnf
failed because of some packages, filesystem being one of the cuplrit.
When listing files and folders in /proc, as shown previously, some files
andf folders belong to nobody:nobody. I can't fix this from my container
(chown root:root --> not permitted) and can't fix it from host as host does
not see what's inside /proc.

**Comment by evverx** [comment]
Well, I understand. This is another issue.

> Do I have any solution?

Try to turn off user namespaces. I.e. you should remove the `-U` flag (see https://github.com/systemd/systemd/commit/af88764ff8fb64f)

Note: you should rebuild and restart all your containers (or run a `fuidshift -r` to map `some-big-number` -> `root` on the host)

**Comment by gabx** [comment]
Removing the flag did not change anything.
I must admit all this is beyond my knowledge. This issue is new (certainly due to a recent update) .

**Comment by gabx** [comment]
Any progress on this issue? I am still left with a container behaving very weirdly because of some directories belonging to `nobody:nobody`. This container is a test server and its current broken state prevents me from working. 
Anything I can do on my side? I fixed some dir with `$ fuidshift`, but it is not enough.

Thank you for any news.

**Comment by mamiu** [comment]
@poettering: This problem also occurs with the `machinectl bind` command.

**Comment by gabx** [comment]
Things have changed a little bit on my setup (Archlinux host running Fedora container). No idea what upgrade (kernel, systemd, Fedora 25...) is causing this change, but UID/GID are now shown differently in the host.
```
# ls -al /var/lib/machines/poppy

dr-xr-xr-x 1 vu-poppy-0 vg-poppy-0   198 Dec 16 17:32 ./
drwx------ 3 root       root        4.0K Dec 15 17:08 ../
dr-xr-xr-x 1 vu-poppy-0 vg-poppy-0     0 Feb  3  2016 boot/
drwxrwxr-x 1 vu-poppy-0 vg-poppy-0    62 Aug 26 19:59 db/
drwxr-xr-x 1 vu-poppy-0 vg-poppy-0     8 Apr 27  2015 dev/
drwxr-xr-x 1 vu-poppy-0 vg-poppy-0  4.2K Feb  3  2016 etc/
drwxr-xr-x 1 vu-poppy-0 vg-poppy-0    86 Feb  3  2016 home/
```


Why this change? How apply now the **shiftuid** command?

**Comment by keszybz** [comment]
@gabx: you have the nss-systemd module which provides user names in the host system for the users and groups internal to the container. That's why `ls` shows those user names.

@poettering I think we should detect when the container has a user namespace (using the same rules as systemd-nspawn), and remap everything to the root user. It'd be *nice* to have an option to remap other users "symbolically", i.e. by name, but that's not easy to implement. Mapping everything to the container's root will solve the OP's problem.

**Comment by poettering** [comment]
Fix waiting in #5333
